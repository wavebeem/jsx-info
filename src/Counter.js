class Counter {
  constructor(options) {
    this._allComponentsCount = 0;
    this._componentCounts = new Map();
    this._componentPropCounts = new Map();
    this._componentChildrenCounts = new Map();
    this._errors = [];
    this._suggestedPlugins = new Set();
    this._sortType = options.sort;
  }

  addParseError({ filename, error }) {
    this._errors.push({ filename, error });
    for (const plugin of error.missingPlugin || []) {
      this._suggestedPlugins.add(plugin);
    }
  }

  incrementComponentUsage(component) {
    const prev = this._componentCounts.get(component) || 0;
    if (!prev) {
      this._componentPropCounts.set(component, new Map());
    }
    this._allComponentsCount++;
    this._componentCounts.set(component, prev + 1);
  }

  incrementComponentPropUsage(component, prop) {
    const map = this._componentPropCounts.get(component);
    const prev = map.get(prop) || 0;
    map.set(prop, prev + 1);
  }

  incrementComponentChildrenUsage(component, childrenNames) {
    if (!this._componentChildrenCounts.has(component)) {
      this._componentChildrenCounts.set(component, new Map());
    }
    const childrenCounts = this._componentChildrenCounts.get(component);
    childrenNames.forEach(child => {
      const prev = childrenCounts.get(child) || 0;
      childrenCounts.set(child, prev + 1);
    });
  }

  getProblems() {
    return {
      errors: [...this._errors],
      suggestedPlugins: [...this._suggestedPlugins]
    };
  }

  getComponentsList() {
    return this.getSortedMap(this._componentCounts).map(x => x.name);
  }

  getAllComponentsReport() {
    const total = this._allComponentsCount;
    const counts = this.getSortedMap(this._componentCounts);
    return { total, counts };
  }

  getComponentChildrenUsageList() {
    return [...this._componentChildrenCounts].sort(
      ([aName, aChildren], [bName, bChildren]) => {
        const sortCount = bChildren.size - aChildren.size;
        const sortName = bName < aName ? -1 : bName > aName ? 1 : 0;
        if (this._sortType === "usage") {
          return sortCount || sortName || 0;
        } else if (this._sortType === "alphabetical") {
          return sortName || sortCount || 0;
        }
        return 0;
      }
    );
  }

  getSortedMap(map) {
    return [...map]
      .map(([name, count]) => {
        return { name, count };
      })
      .sort((a, b) => {
        const sortCount = b.count - a.count;
        const sortName = b.name < a.name ? -1 : b.name > a.name ? 1 : 0;

        if (this._sortType === "usage") {
          return sortCount || sortName || 0;
        } else if (this._sortType === "alphabetical") {
          return sortName || sortCount || 0;
        }
        return 0;
      });
  }

  getComponentReport(component) {
    const componentCount = this._componentCounts.get(component);
    const propCounts = this.getSortedMap(
      this._componentPropCounts.get(component)
    );
    return { componentCount, propCounts };
  }
}

module.exports = Counter;
