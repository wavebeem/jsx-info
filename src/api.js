const parse = require("./parse");
const Reporter = require("./reporter");
const codeSource = require("./code-source");
const { formatPrettyCode } = require("./formatPrettyCode");
const { sleep } = require("./sleep");

async function analyze({
  babelPlugins = [],
  components,
  directory,
  files = ["**/*.{js,jsx,tsx}"],
  gitignore = true,
  ignore = [],
  onFile,
  onStart,
  prop,
  report = ["usage", "props"],
  sort = "usage",
} = {}) {
  if (!prop && report.includes("lines")) {
    throw new Error("`prop` option required for `lines` report");
  }
  const timeStart = Date.now();
  if (onStart) {
    await onStart();
  }
  const filenames = await codeSource.searchForFiles({
    patterns: files,
    gitignore,
    directory,
    ignore,
  });
  const reporter = new Reporter({ sortType: sort });
  for (const filename of filenames) {
    if (onFile) {
      await onFile(filename);
    } else {
      await sleep();
    }
    try {
      const code = codeSource.codeFromFile(filename);
      parse(code, {
        babelPlugins,
        typescript: filename.endsWith(".tsx") || filename.endsWith(".ts"),
        onlyComponents: components,
        onComponent: (componentName) => {
          reporter.addComponent(componentName);
        },
        onProp: ({
          componentName,
          propName,
          propCode,
          startLoc,
          endLoc,
          propValue,
        }) => {
          if (prop) {
            let wantPropKey = prop;
            let wantPropValue = undefined;
            if (prop.includes("=")) {
              const index = prop.indexOf("=");
              const key = prop.slice(0, index);
              const val = prop.slice(index + 1);
              wantPropKey = key;
              wantPropValue = val;
            }
            if (propName !== wantPropKey) {
              return;
            }
            if (wantPropValue !== undefined && propValue !== wantPropValue) {
              return;
            }
          }
          const prettyCode = formatPrettyCode(code, startLoc.line, endLoc.line);
          reporter.addProp({
            componentName,
            propName,
            propCode,
            startLoc,
            endLoc,
            prettyCode,
            filename,
          });
        },
      });
    } catch (error) {
      if (error instanceof SyntaxError) {
        reporter.addParseError(filename, error);
      } else {
        throw error;
      }
    }
  }
  const elapsedTime = (Date.now() - timeStart) / 1000;
  return {
    filenames,
    totals: reporter.getTotals(),
    componentUsage: reporter.getComponentUsage(),
    propUsage: reporter.getPropUsage(),
    lineUsage: reporter.getLineUsage(),
    errors: reporter.getErrors(),
    suggestedPlugins: reporter.getSuggestedPlugins(),
    elapsedTime: elapsedTime,
  };
}

exports.analyze = analyze;
