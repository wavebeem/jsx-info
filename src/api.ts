import { parse as babelParse, ParserPlugin } from "@babel/parser";
import traverse from "@babel/traverse";
import { JSXElement, Node } from "@babel/types";
import fs from "fs";
import globby from "globby";
import path from "path";
import { sleep } from "./sleep";

/** Options passed to `analyze` */
export interface AnalyzeOptions {
  /** List of Babel plugins needed to parse your code */
  babelPlugins?: ParserPlugin[];
  /** Which JSX components should we scan */
  components?: string[];
  /** Which directory should we scan (defaults to current working directory) */
  directory?: string;
  /** Array of file globs (defaults to all .js, .jsx, .tsx files) */
  files?: string[];
  /** Ignore files specified in the .gitignore file (defaults to true) */
  gitignore?: boolean;
  /** Array of file globs to ignore */
  ignore?: string[];
  /** Async callback called before scanning each new file */
  onFile?: (filename: string) => Promise<void>;
  /** Which JSX prop should we scan for (e.g. `id` or `variant=primary`) */
  prop?: string;
  /** Use relative paths instead of absolute paths */
  relativePaths?: boolean;
}

/** A location in source code */
export interface SourceLocation {
  line: number;
  column: number;
}

/** Information about a parse error */
export interface ErrorInfo {
  message: string;
  pos: number;
  loc: SourceLocation;
  missingPlugin: string[];
}

/** Information about a line where a prop was used */
export interface LineInfo {
  propCode: string;
  prettyCode: string;
  startLoc: SourceLocation;
  endLoc: SourceLocation;
  filename: string;
}

export type PropUsage = Record<string, Record<string, number>>;
export type LineUsage = Record<string, Record<string, LineInfo[]>>;
export type ComponentUsage = Record<string, number>;

export interface Analysis {
  filenames: string[];
  componentTotal: number;
  componentUsageTotal: number;
  componentUsage: ComponentUsage;
  propUsage: PropUsage;
  lineUsage: LineUsage;
  errors: Record<string, ErrorInfo>;
  suggestedPlugins: string[];
  elapsedTime: number;
  directory: string;
}

/**
 * Return a Promise with the JSX usage analysis of a project
 *
 * This function can easily return several megabytes of data. Parsing every
 * single JS/TS file in a project can take a while. Please be patient. The
 * `onFile` callbacks are async. Parsing each file is not an async action and
 * may block for a couple seconds if the file is large. You can insert a "sleep"
 * command into the `onFile` callback if you would like to return control
 * briefly.
 */
export async function analyze({
  babelPlugins = [],
  components,
  directory = process.cwd(),
  files = ["**/*.{js,jsx,tsx}"],
  gitignore = true,
  ignore = [],
  onFile,
  prop = "",
  relativePaths = false,
}: AnalyzeOptions): Promise<Analysis> {
  function processFilename(filename: string): string {
    if (relativePaths) {
      return filename;
    }
    return path.resolve(directory, filename);
  }
  const searchProp = prop;
  const timeStart = Date.now();
  const filenames = await globby(files || "**/*.{js,jsx,tsx}", {
    absolute: !relativePaths,
    onlyFiles: true,
    gitignore,
    ignore,
    cwd: directory,
  });
  const reporter = new Reporter();
  for (const filename of filenames) {
    if (onFile) {
      await onFile(processFilename(filename));
    } else {
      await sleep();
    }
    try {
      const code = fs.readFileSync(path.resolve(directory, filename), "utf8");
      parse(code, {
        babelPlugins,
        typescript: filename.endsWith(".tsx") || filename.endsWith(".ts"),
        onlyComponents: components,
        onComponent: ({ componentName, node }) => {
          reporter.addComponent(componentName);
          const props = node.openingElement.attributes.map((propNode) => {
            if (!propNode.loc) {
              throw new Error(`JSXElement propNode.loc is missing`);
            }
            return {
              componentName,
              propName: createProp(propNode),
              propCode: code.slice(propNode.start || 0, propNode.end || -1),
              startLoc: propNode.loc.start,
              endLoc: propNode.loc.end,
              propValue:
                propNode.type === "JSXAttribute"
                  ? formatPropValue(propNode.value)
                  : formatPropValue(null),
            };
          });
          if (searchProp.startsWith("!")) {
            const wantProp = searchProp.slice(1);
            if (props.every((p) => p.propName !== wantProp)) {
              if (!node.loc) {
                throw new Error(`JSXElement propNode.loc is missing`);
              }
              reporter.addProp(componentName, searchProp, {
                propCode: code.slice(node.start || 0, node.end || -1),
                startLoc: node.loc.start,
                endLoc: node.loc.end,
                prettyCode: formatPrettyCode(
                  code,
                  node.loc.start.line,
                  node.loc.end.line
                ),
                filename: processFilename(filename),
              });
            }
          } else {
            for (const prop of props) {
              let wantPropKey = searchProp;
              let match = (_value: string | symbol): boolean => true;
              if (searchProp.includes("!=")) {
                const index = searchProp.indexOf("!=");
                const key = searchProp.slice(0, index);
                const val = searchProp.slice(index + 2);
                wantPropKey = key;
                match = (value) => value !== val;
              } else if (searchProp.includes("=")) {
                const index = searchProp.indexOf("=");
                const key = searchProp.slice(0, index);
                const val = searchProp.slice(index + 1);
                wantPropKey = key;
                match = (value) => value === val;
              }
              if (wantPropKey && prop.propName !== wantPropKey) {
                return;
              }
              if (!match(prop.propValue)) {
                return;
              }
              reporter.addProp(componentName, prop.propName, {
                propCode: prop.propCode,
                startLoc: prop.startLoc,
                endLoc: prop.endLoc,
                prettyCode: formatPrettyCode(
                  code,
                  prop.startLoc.line,
                  prop.endLoc.line
                ),
                filename: processFilename(filename),
              });
            }
          }
        },
      });
    } catch (error) {
      if (error instanceof SyntaxError) {
        reporter.addParseError(processFilename(filename), error);
      } else {
        throw error;
      }
    }
  }
  const elapsedTime = (Date.now() - timeStart) / 1000;
  return {
    filenames,
    componentTotal: reporter.getComponentTotal(),
    componentUsageTotal: reporter.getComponentUsageTotal(),
    componentUsage: reporter.components,
    propUsage: reporter.props,
    lineUsage: reporter.lines,
    errors: reporter.errors,
    suggestedPlugins: reporter.suggestedPlugins,
    elapsedTime: elapsedTime,
    directory,
  };
}

////////////////////////////////////////////////////////////////////////////////
//
// Implementation details below
//
////////////////////////////////////////////////////////////////////////////////

const linesCache = new Map<string, string[]>();

function getLines(code: string): string[] {
  let lines = linesCache.get(code);
  if (lines) {
    return lines;
  }
  lines = code.split(/\r?\n/);
  linesCache.set(code, lines);
  return lines;
}

const EXPRESSION = Symbol("formatPropValue.EXPRESSION");

function formatPropValue(value: Node | null): string | symbol {
  if (value === null) {
    return "true";
  }
  if (!value) {
    return EXPRESSION;
  }
  switch (value.type) {
    // TODO: Should we interpret anything else here?
    case "StringLiteral":
      return value.value;
    case "JSXExpressionContainer":
      return formatPropValue(value.expression);
    case "NumericLiteral":
    case "BooleanLiteral":
      return String(value.value);
    default:
      return EXPRESSION;
  }
}

function getAttributeName(attributeNode: Node): string {
  switch (attributeNode.type) {
    case "JSXIdentifier":
      return attributeNode.name;
    case "JSXNamespacedName":
      return attributeNode.name.name;
    case "JSXAttribute":
      return getAttributeName(attributeNode.name);
    case "JSXSpreadAttribute":
      return "{...}";
    default:
      throw new Error(`unexpected node type: ${attributeNode.type}`);
  }
}

function createProp(attributeNode: Node) {
  return getAttributeName(attributeNode);
}

function getDottedName(nameNode: Node): string {
  switch (nameNode.type) {
    case "JSXMemberExpression":
      return [nameNode.object, nameNode.property].map(getDottedName).join(".");
    case "JSXIdentifier":
      return nameNode.name;
    default:
      throw new Error(`unexpected node type: ${nameNode.type}`);
  }
}

function createComponent(componentNode: JSXElement) {
  return getDottedName(componentNode.openingElement.name);
}

interface ParseOptions {
  typescript?: boolean;
  babelPlugins?: ParserPlugin[];
  onlyComponents?: string[];
  onComponent?: (options: { componentName: string; node: JSXElement }) => void;
}

export function parse(code: string, options: ParseOptions = {}): void {
  const {
    typescript = false,
    babelPlugins = [],
    onlyComponents = [],
    onComponent = () => {},
  } = options;
  function doReportComponent(component: string) {
    if (onlyComponents.length === 0) {
      return true;
    }
    return onlyComponents.indexOf(component) !== -1;
  }
  const ast = babelParse(code, {
    sourceType: "unambiguous",
    allowReturnOutsideFunction: true,
    plugins: [
      typescript ? "typescript" : "flow",
      "jsx",
      "dynamicImport",
      "classProperties",
      "objectRestSpread",
      ...babelPlugins,
    ],
  });
  traverse(ast, {
    JSXElement(path) {
      const node = path.node;
      const componentName = createComponent(node);
      if (doReportComponent(componentName)) {
        onComponent({ componentName, node });
      }
    },
  });
}

function formatPrettyCode(
  code: string,
  startLine: number,
  endLine: number
): string {
  const output: string[] = [];
  const lines = getLines(code);
  // Line numbers should be padded to at least 4 digits for consistency and
  // readability, but let's also let them grow if we have super long files :|
  const maxDigits = Math.max(
    String(startLine).length,
    String(endLine).length,
    4
  );
  for (let lineno = startLine; lineno <= endLine; lineno++) {
    output.push(String(lineno).padStart(maxDigits) + " | " + lines[lineno - 1]);
  }
  return output.join("\n");
}

class Reporter {
  components: ComponentUsage = {};
  props: PropUsage = {};
  lines: LineUsage = {};
  suggestedPlugins: string[] = [];
  errors: Record<string, ErrorInfo> = {};

  // TODO: Does Babel expose the actual error type for us?
  addParseError(filename: string, error: any): void {
    this.errors[filename] = {
      message: error.message,
      pos: error.pos,
      loc: error.loc,
      missingPlugin: error.missingPlugin,
    };
    for (const plugin of error.missingPlugin || []) {
      if (!this.suggestedPlugins.includes(plugin)) {
        this.suggestedPlugins.push(plugin);
      }
    }
  }

  addComponent(componentName: string): void {
    this.components[componentName] = (this.components[componentName] || 0) + 1;
  }

  private _incrementProp(componentName: string, propName: string): void {
    if (!this.props[componentName]) {
      this.props[componentName] = {};
    }
    const count = this.props[componentName][propName] || 0;
    this.props[componentName][propName] = count + 1;
  }

  private _ensureLines(componentName: string, propName: string): void {
    if (!this.lines[componentName]) {
      this.lines[componentName] = {};
    }
    if (!this.lines[componentName][propName]) {
      this.lines[componentName][propName] = [];
    }
  }

  addProp(componentName: string, propName: string, line: LineInfo) {
    this._incrementProp(componentName, propName);
    this._ensureLines(componentName, propName);
    this.lines[componentName][propName].push(line);
  }

  getComponentTotal(): number {
    return Object.keys(this.components).length;
  }

  getComponentUsageTotal(): number {
    return Object.values(this.components).reduce((a, b) => a + b, 0);
  }
}
