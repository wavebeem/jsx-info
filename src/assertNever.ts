export function assertNever(value: never): never {
  throw new Error(`assertNever: ${value}`);
}
