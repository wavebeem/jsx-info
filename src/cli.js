const program = require("commander");
const pkg = require("../package.json");

function listOption(x, acc = []) {
  acc.push(x);
  return acc;
}

program.name(pkg.name);
program.version(pkg.version, "-v, --version");
program
  .arguments("[components...]")
  .description("Displays a report of JSX component and prop usage")
  .option("--no-progress", "do not show progress messages")
  // NOTE: chalk automatically reads process.argv for these flags, and commander
  // does not correctly handle allowing both "--foo" and "--no-foo" flags
  //
  // https://github.com/tj/commander.js/issues/108
  .option("--color", "force enable color")
  .option("--no-color", "force disable color")
  .option("--no-gitignore", "disable reading .gitignore files")
  .option(
    "--add-babel-plugin <plugin>",
    "adds a babel plugin (repeatable)",
    listOption
  )
  .option(
    "--directory <directory>",
    "directory to use as the base for finding files instead of cwd"
  )
  .option(
    "--ignore <pattern>",
    "adds a glob pattern used to ignore input files (repeatable)",
    listOption
  )
  .option(
    "--files <pattern>",
    "glob pattern used to find input files",
    "**/*.{js,jsx,tsx}"
  )
  .option(
    "--sort <alphabetical|usage>",
    "specify sort type of the report",
    /^(alphabetical|usage)$/i,
    "usage"
  )
  .option(
    "--report <usage|props>",
    "specify reports to show (can be repeted)",
    listOption
  )
  .parse(process.argv);

program.on("--help", () => {
  // eslint-disable-next-line no-console
  console.log(`
Examples:
  # Display info for every component
  $ ${pkg.name}

  # Display info only for <div> and <Tab.Container>
  $ ${pkg.name} div Tab.Container


  # Ignore any folder named at any depth named \`__test__\`,
  # as well as \`packages/legacy\`
  $ ${pkg.name} --ignore '**/__test__' --ignore packages/legacy


  # Enable Babel plugins
  $ ${
    pkg.name
  } --add-babel-plugin decorators-legacy --add-babel-plugin pipelineOperator

Documentation can be found at https://github.com/wavebeem/jsx-info
`);
});

module.exports = {
  components: program.args,
  showProgress: program.progress,
  babelPlugins: program.addBabelPlugin || [process.cwd()],
  directory: program.directory,
  gitignore: program.gitignore,
  ignore: program.ignore || [],
  files: program.files,
  sort: program.sort,
  report: program.report || []
};
