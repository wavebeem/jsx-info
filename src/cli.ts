import { ParserPlugin } from "@babel/parser";
import chalk from "chalk";
import program from "commander";
import { cosmiconfigSync } from "cosmiconfig";
import path from "path";
import { version } from "../package.json";
import { print, printError } from "./printer";

export type ReportType = "usage" | "props" | "lines";

program.name("jsx-info");
program.version(version, "-v, --version");
program
  .description("Displays a report of JSX component and prop usage")
  .option("--no-progress", "don't show progress messages")
  .option("--components <components...>", "which components to scan")
  .option("--color", "force enable color")
  .option("--no-color", "force disable color")
  .option("--no-config", "disable config file")
  .option("--no-gitignore", "disable reading .gitignore files")
  .option("--babel-plugins <plugins...>", "set Babel plugins")
  .option(
    "--directory <directory>",
    "directory to use as the base for finding files"
  )
  .option("--ignore <patterns...>", "glob patterns used to ignore input files")
  .option("--files <patterns...>", "glob patterns used to find input files")
  .option("--report <usage|props|lines>", "which report to show")
  .option(
    "--prop <prop[=value]>",
    "which prop to search for when running a lines report"
  );

program.on("--help", () => {
  print(chalk`
Examples:
  Use jsx-info interactively:
      {bold.cyan npx jsx-info}

  Display info only for <div> and <Tab.Container>
      {bold.cyan npx jsx-info --components} {underline.magenta div Tab.Container}

  See lines where className prop was used on <div> component
      {bold.cyan npx jsx-info --report} {underline.magenta lines} {bold.cyan --prop} {underline.magenta className} {bold.cyan --components} {underline.magenta div}

  See lines where \`id\` prop was used on any component
      {bold.cyan npx jsx-info --report} {underline.magenta lines} {bold.cyan --prop} {underline.magenta id} {bold.cyan --components} {underline.magenta "*"}

  See lines where kind prop was used with value "primary" on Button component
      {bold.cyan npx jsx-info --report} {underline.magenta lines} {bold.cyan --prop} {underline.magenta kind=primary} {bold.cyan --components} {underline.magenta Button}

  Ignore folders based on glob paths
      {bold.cyan npx jsx-info --ignore} {underline.magenta "**/__test__"} {underline.magenta packages/legacy}

  Enable Babel plugins
      {bold.cyan npx jsx-info --babel-plugins} {underline.magenta decorators-legacy} {underline.magenta pipelineOperator}

  Example .jsx-info.json config file
      ${"{"}
        {cyan "babelPlugins"}: [{magenta "decorators-legacy"}, {magenta "pipelineOperator"}],
        {cyan "directory"}: {magenta "src"},
        {cyan "ignore"}: [{magenta "**/__test__"}, {magenta "legacy/**"}],
        {cyan "files"}: [{magenta "**/*.${"{"}js,jsx,tsx${"}"}"}]
      ${"}"}
`);
});

program.parse(process.argv);

print(chalk.magenta.bold(`jsx-info ${version} by @wavebeem`));

function getConfig() {
  if (!program.config) {
    return {};
  }
  try {
    const explorer = cosmiconfigSync("jsx-info", {
      searchPlaces: ["package.json", ".jsx-info.json"],
    });
    const result = explorer.search();
    if (result) {
      print(`\nLoaded configuration from ${result.filepath}\n`);
      if (result.config.directory) {
        result.config.directory = path.resolve(
          result.filepath,
          "..",
          result.config.directory
        );
      }
      return result.config;
    }
  } catch (err) {
    if (process.env.DEBUG) {
      // eslint-disable-next-line no-console
      console.log(err.stack);
    }
    printError("failed to parse config file");
  }
  return {};
}

const config = getConfig();

function concat<T>(a: T[] = [], b: T[] = []): T[] {
  return [...a, ...b];
}

export const prop: string = program.prop;
export const components: string[] = program.components;
export const showProgress: boolean = program.progress;
export const babelPlugins: ParserPlugin[] = concat(
  config.babelPlugins,
  program.babelPlugins
);
export const directory: string = program.directory || config.directory;
export const gitignore: boolean = program.gitignore;
export const ignore: string[] = concat(program.ignore, config.ignore);
export const files: string[] = concat<string>(program.files, config.files);
export const report: ReportType = program.report;
