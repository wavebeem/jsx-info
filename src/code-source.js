const globby = require("globby");
const fs = require("fs");

const codeFromFile = filename => fs.readFileSync(filename, "utf8");

const searchForFiles = async ({ patterns, gitignore, directory, ignore }) => {
  return await globby(patterns, {
    absolute: true,
    onlyFiles: true,
    gitignore,
    ignore,
    cwd: directory || process.cwd()
  });
};

exports.codeFromFile = codeFromFile;
exports.searchForFiles = searchForFiles;
