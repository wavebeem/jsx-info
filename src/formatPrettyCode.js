const printer = require("./printer");

const cache = new Map();

function getLines(code) {
  let lines = cache.get(code);
  if (lines) {
    return lines;
  }
  lines = code.split(/\r?\n/);
  cache.set(code, lines);
  return lines;
}

function formatPrettyCode(code, startLine, endLine) {
  const output = [];
  const lines = getLines(code);
  // Line numbers should be padded to at least 4 digits for consistency and
  // readability, but let's also let them grow if we have super long files :|
  const maxDigits = Math.max(
    String(startLine).length,
    String(endLine).length,
    4
  );
  for (let lineno = startLine; lineno <= endLine; lineno++) {
    output.push(
      printer.styleLinenos(String(lineno).padStart(maxDigits) + " | ") +
        lines[lineno - 1]
    );
  }
  return output.join("\n");
}

exports.formatPrettyCode = formatPrettyCode;
