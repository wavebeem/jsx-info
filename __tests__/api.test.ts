import * as path from "path";
import { analyze } from "../src/api";

test("analyze empty directory", async () => {
  const directory = path.resolve(__dirname, "../fixtures/empty");
  const analysis = await analyze({ directory, relativePaths: true });
  expect(analysis).toMatchSnapshot({
    directory: expect.any(String),
    elapsedTime: expect.any(Number),
  });
});

test("analyze basic directory with absolute paths", async () => {
  const directory = path.resolve(__dirname, "../fixtures/basic");
  const analysis = await analyze({ directory });
  expect(analysis.filenames).toMatchObject([
    path.resolve(directory, "main.js"),
  ]);
});

test("analyze basic directory", async () => {
  const directory = path.resolve(__dirname, "../fixtures/basic");
  const analysis = await analyze({ directory, relativePaths: true });
  expect(analysis).toMatchSnapshot({
    directory: expect.any(String),
    elapsedTime: expect.any(Number),
  });
});

test("analyze ts directory", async () => {
  const directory = path.resolve(__dirname, "../fixtures/ts");
  const analysis = await analyze({ directory, relativePaths: true });
  expect(analysis).toMatchSnapshot({
    directory: expect.any(String),
    elapsedTime: expect.any(Number),
  });
});

test("analyze error directory", async () => {
  const directory = path.resolve(__dirname, "../fixtures/error");
  const analysis = await analyze({ directory, relativePaths: true });
  expect(analysis).toMatchSnapshot({
    directory: expect.any(String),
    elapsedTime: expect.any(Number),
  });
});

test("analyze decorators directory", async () => {
  const directory = path.resolve(__dirname, "../fixtures/decorators");
  const analysis = await analyze({ directory, relativePaths: true });
  expect(analysis).toMatchSnapshot({
    directory: expect.any(String),
    elapsedTime: expect.any(Number),
  });
});

test("analyze decorators directory with plugin decorators-legacy", async () => {
  const directory = path.resolve(__dirname, "../fixtures/decorators");
  const analysis = await analyze({
    directory,
    babelPlugins: ["decorators-legacy"],
    relativePaths: true,
  });
  expect(analysis).toMatchSnapshot({
    directory: expect.any(String),
    elapsedTime: expect.any(Number),
  });
});

test("analyze prop-usage directory", async () => {
  const directory = path.resolve(__dirname, "../fixtures/prop-usage");
  const analysis = await analyze({
    directory,
    components: ["div"],
    prop: "a",
    relativePaths: true,
  });
  expect(analysis).toMatchSnapshot({
    directory: expect.any(String),
    elapsedTime: expect.any(Number),
  });
});
