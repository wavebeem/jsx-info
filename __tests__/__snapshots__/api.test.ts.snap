// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze basic directory 1`] = `
Object {
  "componentTotal": 1,
  "componentUsage": Object {
    "div": 1,
  },
  "componentUsageTotal": 1,
  "directory": Any<String>,
  "elapsedTime": Any<Number>,
  "errors": Object {},
  "filenames": Array [
    "main.js",
  ],
  "lineUsage": Object {
    "div": Object {
      "a": Array [
        Object {
          "endLoc": Position {
            "column": 6,
            "line": 1,
          },
          "filename": "main.js",
          "prettyCode": "   1 | <div a b c />;",
          "propCode": "a",
          "startLoc": Position {
            "column": 5,
            "line": 1,
          },
        },
      ],
      "b": Array [
        Object {
          "endLoc": Position {
            "column": 8,
            "line": 1,
          },
          "filename": "main.js",
          "prettyCode": "   1 | <div a b c />;",
          "propCode": "b",
          "startLoc": Position {
            "column": 7,
            "line": 1,
          },
        },
      ],
      "c": Array [
        Object {
          "endLoc": Position {
            "column": 10,
            "line": 1,
          },
          "filename": "main.js",
          "prettyCode": "   1 | <div a b c />;",
          "propCode": "c",
          "startLoc": Position {
            "column": 9,
            "line": 1,
          },
        },
      ],
    },
  },
  "propUsage": Object {
    "div": Object {
      "a": 1,
      "b": 1,
      "c": 1,
    },
  },
  "suggestedPlugins": Array [],
}
`;

exports[`analyze decorators directory 1`] = `
Object {
  "componentTotal": 0,
  "componentUsage": Object {},
  "componentUsageTotal": 0,
  "directory": Any<String>,
  "elapsedTime": Any<Number>,
  "errors": Object {
    "main.js": Object {
      "loc": Position {
        "column": 0,
        "line": 1,
      },
      "message": "This experimental syntax requires enabling one of the following parser plugin(s): 'decorators-legacy, decorators' (1:0)",
      "missingPlugin": Array [
        "decorators-legacy",
        "decorators",
      ],
      "pos": 0,
    },
  },
  "filenames": Array [
    "main.js",
  ],
  "lineUsage": Object {},
  "propUsage": Object {},
  "suggestedPlugins": Array [
    "decorators-legacy",
    "decorators",
  ],
}
`;

exports[`analyze decorators directory with plugin decorators-legacy 1`] = `
Object {
  "componentTotal": 2,
  "componentUsage": Object {
    "App": 1,
    "div": 1,
  },
  "componentUsageTotal": 2,
  "directory": Any<String>,
  "elapsedTime": Any<Number>,
  "errors": Object {},
  "filenames": Array [
    "main.js",
  ],
  "lineUsage": Object {},
  "propUsage": Object {},
  "suggestedPlugins": Array [],
}
`;

exports[`analyze empty directory 1`] = `
Object {
  "componentTotal": 0,
  "componentUsage": Object {},
  "componentUsageTotal": 0,
  "directory": Any<String>,
  "elapsedTime": Any<Number>,
  "errors": Object {},
  "filenames": Array [],
  "lineUsage": Object {},
  "propUsage": Object {},
  "suggestedPlugins": Array [],
}
`;

exports[`analyze error directory 1`] = `
Object {
  "componentTotal": 0,
  "componentUsage": Object {},
  "componentUsageTotal": 0,
  "directory": Any<String>,
  "elapsedTime": Any<Number>,
  "errors": Object {
    "main.js": Object {
      "loc": Position {
        "column": 0,
        "line": 2,
      },
      "message": "Unexpected token (2:0)",
      "missingPlugin": undefined,
      "pos": 12,
    },
  },
  "filenames": Array [
    "main.js",
  ],
  "lineUsage": Object {},
  "propUsage": Object {},
  "suggestedPlugins": Array [],
}
`;

exports[`analyze prop-usage directory 1`] = `
Object {
  "componentTotal": 1,
  "componentUsage": Object {
    "div": 8,
  },
  "componentUsageTotal": 8,
  "directory": Any<String>,
  "elapsedTime": Any<Number>,
  "errors": Object {},
  "filenames": Array [
    "main.js",
  ],
  "lineUsage": Object {
    "div": Object {
      "a": Array [
        Object {
          "endLoc": Position {
            "column": 6,
            "line": 1,
          },
          "filename": "main.js",
          "prettyCode": "   1 | <div a />;",
          "propCode": "a",
          "startLoc": Position {
            "column": 5,
            "line": 1,
          },
        },
        Object {
          "endLoc": Position {
            "column": 6,
            "line": 3,
          },
          "filename": "main.js",
          "prettyCode": "   3 | <div a b />;",
          "propCode": "a",
          "startLoc": Position {
            "column": 5,
            "line": 3,
          },
        },
        Object {
          "endLoc": Position {
            "column": 6,
            "line": 4,
          },
          "filename": "main.js",
          "prettyCode": "   4 | <div a b c />;",
          "propCode": "a",
          "startLoc": Position {
            "column": 5,
            "line": 4,
          },
        },
      ],
    },
  },
  "propUsage": Object {
    "div": Object {
      "a": 3,
    },
  },
  "suggestedPlugins": Array [],
}
`;

exports[`analyze ts directory 1`] = `
Object {
  "componentTotal": 1,
  "componentUsage": Object {
    "div": 1,
  },
  "componentUsageTotal": 1,
  "directory": Any<String>,
  "elapsedTime": Any<Number>,
  "errors": Object {},
  "filenames": Array [
    "main.tsx",
  ],
  "lineUsage": Object {
    "div": Object {
      "typescriptIsCool": Array [
        Object {
          "endLoc": Position {
            "column": 21,
            "line": 1,
          },
          "filename": "main.tsx",
          "prettyCode": "   1 | <div typescriptIsCool />;",
          "propCode": "typescriptIsCool",
          "startLoc": Position {
            "column": 5,
            "line": 1,
          },
        },
      ],
    },
  },
  "propUsage": Object {
    "div": Object {
      "typescriptIsCool": 1,
    },
  },
  "suggestedPlugins": Array [],
}
`;
